commit ca6b0bf0e086513b9ee5efc0aa5770ecb57778af
Author: Al Viro <viro@zeniv.linux.org.uk>
Date:   Tue Feb 15 22:04:37 2011 -0500

    pagemap: close races with suid execve
    
    just use mm_for_maps()
    
    Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
    [dannf: backported to Debian's 2.6.26]

commit 4fb7cdfbc27b0635a9ec66200291d2d2babb9970
Author: dann frazier <dannf@debian.org>
Date:   Wed Aug 31 22:06:29 2011 -0600

    bugfix/all/pagemap-close-races-with-suid-execve.patch

diff --git a/fs/proc/base.c b/fs/proc/base.c
index bce2890..47afca0 100644
--- a/fs/proc/base.c
+++ b/fs/proc/base.c
@@ -2440,7 +2440,7 @@ static const struct pid_entry tgid_base_stuff[] = {
 #ifdef CONFIG_PROC_PAGE_MONITOR
 	REG("clear_refs", S_IWUSR, clear_refs),
 	REG("smaps",      S_IRUGO, smaps),
-	REG("pagemap",    S_IRUSR, pagemap),
+	REG("pagemap",    S_IRUGO, pagemap),
 #endif
 #ifdef CONFIG_SECURITY
 	DIR("attr",       S_IRUGO|S_IXUGO, attr_dir),
@@ -2776,7 +2776,7 @@ static const struct pid_entry tid_base_stuff[] = {
 #ifdef CONFIG_PROC_PAGE_MONITOR
 	REG("clear_refs", S_IWUSR, clear_refs),
 	REG("smaps",     S_IRUGO, smaps),
-	REG("pagemap",    S_IRUSR, pagemap),
+	REG("pagemap",    S_IRUGO, pagemap),
 #endif
 #ifdef CONFIG_SECURITY
 	DIR("attr",      S_IRUGO|S_IXUGO, attr_dir),
diff --git a/fs/proc/task_mmu.c b/fs/proc/task_mmu.c
index 8feda82..56c00dc 100644
--- a/fs/proc/task_mmu.c
+++ b/fs/proc/task_mmu.c
@@ -663,7 +663,8 @@ static ssize_t pagemap_read(struct file *file, char __user *buf,
 		goto out;
 
 	ret = -EACCES;
-	if (!ptrace_may_attach(task))
+	mm = mm_for_maps(task);
+	if (!mm)
 		goto out_task;
 
 	ret = -EINVAL;
@@ -672,10 +673,6 @@ static ssize_t pagemap_read(struct file *file, char __user *buf,
 		goto out_task;
 
 	ret = 0;
-	mm = get_task_mm(task);
-	if (!mm)
-		goto out_task;
-
 
 	uaddr = (unsigned long)buf & PAGE_MASK;
 	uend = (unsigned long)(buf + count);
